### PR TITLE
[ObjectStore] Add `append` API impl for `LocalFileSystem`

### LINKED ISSUE #3740: Support for Async CSV Writer
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

I want an async CSV writer to use it in an async context. The current implementation is blocking even if we are in the tokio environment.

**Describe the solution you'd like**
A writer using `tokio::io::AsyncWrite` might be the solution.

**Describe alternatives you've considered**
NA

**Additional context**
Na

### LINKED ISSUE #3742: Support for Async JSON Writer
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

I want an async JSON writer to use it in an async context. The current implementation is blocking even if we are in the tokio environment.

**Describe the solution you'd like**
A writer using `tokio::io::AsyncWrite` might be the solution.

**Describe alternatives you've considered**
NA

**Additional context**
Na

### PR DESCRIPTION
**Pull request by metesynnada**
# Which issue does this PR close?

Closes #3740 and closes #3742 since we go down a different path.

# Rationale for this change

This PR adds the `LocalFileSystem` implementation for the "append" API. 

# What changes are included in this PR?

- Append API implementation for `LocalFileSystem`.

However, to simplify things, we used the `tokio::fs` feature which is not supported on `wasm32`. There is another implementation on https://github.com/synnada-ai/arrow-rs/pull/5 that alters the `LocalUpload` struct to support both staged and unstaged file writings.

# Are there any user-facing changes?

No.

### CODE REVIEW

**Review by alamb — COMMENTED** [review]
Looks good to me -- thank you @metesynnada 

I also verified that the semantics match what is described on https://github.com/apache/arrow-rs/blob/fb35d264225c9aa7aa5eeefbca72e914590fac25/object_store/src/lib.rs#L315-L334

cc @tustvold

File: arrow-csv/src/writer.rs [hunk 1]
```diff
@@ -194,7 +194,7 @@ impl<W: Write> Writer<W> {
 }
 
 /// A CSV writer builder
-#[derive(Debug)]
+#[derive(Clone, Debug)]
```
Comment (alamb) [inline arrow-csv/src/writer.rs:197]: 👍

**Review by alamb — APPROVED** [review]

**Review by tustvold — COMMENTED** [review]
My experience of target-specific feature selection has been that they cause very strange and hard to debug problems, the ahash getrandom nonsense is a frequent source of pain.

I wonder if we could instead add a default-disabled top-level feature, maybe `append_local` that gates this functionality? This would also help with flagging the functionality as not part of the standard object store contract.

**Comment by ozankabak** [comment]
IMO the target-specific selection in this case is straightforward/surgical and I don't immediately see any potential headaches related to this in the future. Given that the only target this doesn't support this is WASM, keeping the selection seems OK to me.

BTW, I am not categorically against the `append_local` idea as it doesn't really matter for Datafusion -- we will simply enable it there. However, it seems like an avoidable complexity in general. Our initial thinking was to remove the selection when support arrives, and we already discovered [workarounds](https://github.com/synnada-ai/arrow-rs/pull/5) in case we want to support it sooner rather than later.

@alamb, @tustvold: Let us know what you think and we can finalize accordingly. Thanks for the reviews!

**Comment by alamb** [comment]
> > My experience of target-specific feature selection has been that they cause very strange and hard to debug problems, the ahash getrandom nonsense is a frequent source of pain.

> I wonder if we could instead add a default-disabled top-level feature, maybe append_local that gates this functionality? This would also help with flagging the functionality as not part of the standard object store contract.

I am not sure about the WASM usecase in general for object_store (as in I don't really understand why it is an  an important target to support for object_store at all). If we are looking for simplifications maybe we can remove that feature support instead of making a new feature flag 🤔 


I agree with @ozankabak that making a new feature flag for append seems overly complicated. I think this would be fine to merge as is. 

What do you think @tustvold ?

**Comment by tustvold** [comment]
Fine with me, was just a suggestion.

**Comment by alamb** [comment]
Thanks everyone@!

**Pull request merged**